src/common/src/util/value_encoding/mod.rs

```diff
@@ -147,69 +168,72 @@ fn deserialize_value(ty: &DataType, mut data: impl Buf) -> Result<Datum> {
         DataType::Timestamp => ScalarImpl::NaiveDateTime(deserialize_naivedatetime(data)?),
         DataType::Timestampz => ScalarImpl::Int64(data.get_i64_le()),
         DataType::Date => ScalarImpl::NaiveDate(deserialize_naivedate(data)?),
-        DataType::Struct { .. } => deserialize_struct(ty, data)?,
-        DataType::List { .. } => deserialize_list(ty, data)?,
+        DataType::Struct(struct_def) => deserialize_struct(struct_def, data)?,
+        DataType::List {
+            datatype: item_type,
+        } => deserialize_list(item_type, data)?,
     }))
 }
 
-fn deserialize_struct(data_type: &DataType, mut data: impl Buf) -> Result<ScalarImpl> {
-    assert_eq!(data_type.prost_type_name(), TypeName::Struct);
-    let len = data.get_u32_le();
-    let mut bytes = vec![0; len as usize];
-    data.copy_to_slice(&mut bytes);
-    ScalarImpl::from_proto_bytes(&bytes, &data_type.to_protobuf())
-        .map_err(ValueEncodingError::InvalidStructEncoding)
+fn deserialize_struct(struct_def: &StructType, data: &mut impl Buf) -> Result<ScalarImpl> {
+    let field_values = struct_def
+        .fields
+        .iter()
+        .map(|field_type| inner_deserialize_datum(data, field_type))
+        .collect::<Result<Vec<Datum>>>()?;
```

Comment:
For example, I think this should work.

```suggestion
fn deserialize_struct(struct_def: &StructType, data: impl Buf) -> Result<ScalarImpl> {
    let field_values = struct_def
        .fields
        .iter()
        .map(|field_type| deserialize_datum(&mut data, field_type))
        .collect::<Result<Vec<Datum>>>()?;
```

Reply:
In fact, that's what I did in the beginning. But this is a recursive invocation and the number of uses of `&mut buf` overflows.
<img width="1508" alt="569e6a660df963bea6a70ac7abc67fd6" src="https://user-images.githubusercontent.com/33933578/196140098-c41335fa-c0e4-44d8-8774-e2ff951a9b53.png">

Reply:
Okay... that makes sense. As the recursion depth of `Struct` scalar is unlimited, then we'll generate unlimited types with this function.

Reply:
Here is `deserialization`, and `iter_elems_ref!` is used for serialization.